nixos/modules/services/networking/wireguard.nix

```diff
@@ -0,0 +1,165 @@
+
+      privateKey = mkOption {
+        example = "yAnz5TF+lXXJte14tji3zlMNq+hd2rYUIgJBgB3fBmk=";
+        type = types.str;
+        description = "Private key used by the interface.";
+      };
+
+      presharedKey = mkOption {
+        default = null;
+        example = "rVXs/Ni9tu3oDBLS4hOyAUAa1qTWVA3loR8eL20os3I=";
+        type = with types; nullOr str;
+        description = "Preshared key key used by the interface.";
+      };
+
+      listenPort = mkOption {
+        default = null;
+        type = with types; nullOr int;
+        example = 12345;
+        description = "Port used by the interface.";
+      };
+
+      peers = mkOption {
+        default = [];
+        description = "Peers linked to the interface.";
+        type = with types; listOf (submodule peerOpts);
+      };
+
+    };
+
+  };
+
+  # peer options
+
+  peerOpts = {
+
+    options = {
+
+      publicKey = mkOption {
+        example = "xTIBA5rboUvnH4htodjb6e697QjLERt1NAB4mZqp8Dg=";
+        type = types.str;
+        description = "Public key of the peer.";
+      };
+
+      allowedIPs = mkOption {
+        example = [ "10.192.122.3/32" "10.192.124.1/24" ];
+        type = with types; listOf str;
+        description = "IPs allowed by the peer.";
+      };
+
+      endpoint = mkOption {
+        default = null;
+        example = "test.wireguard.io:18981";
+        type = with types; nullOr str;
+        description = "Endpoint of the peer.";
+      };
+
+      persistentKeepalive = mkOption {
+        default = null;
+        type = with types; nullOr int;
+        example = 12345;
+        description = "Keepalive seconds.";
+      };
+
+    };
+
+  };
+
+  generateConf = name: values: pkgs.writeText "wireguard-${name}.conf" ''
+    [Interface]
+    PrivateKey = ${values.privateKey}
+    ${optionalString (values.presharedKey != null) "PresharedKey = ${values.presharedKey}"}
+    ${optionalString (values.listenPort != null)   "ListenPort = ${toString values.listenPort}"}
+  
+    ${concatStringsSep "\n\n" (map (peer: ''
+    [Peer]
+    PublicKey = ${peer.publicKey}
+    ${optionalString (peer.allowedIPs != []) "AllowedIPs = ${concatStringsSep ", " peer.allowedIPs}"}
+    ${optionalString (peer.endpoint != null) "Endpoint = ${peer.endpoint}"}
+    ${optionalString (peer.persistentKeepalive != null) "PersistentKeepalive = ${toString peer.persistentKeepalive}"}
+    '') values.peers)}
+  '';
+
+  generateUnit = name: values: 
+    nameValuePair "wireguard-${name}" 
+      {
+        description = "Wireguard unit for interface ${name}";
+        wantedBy = [ "ip-up.target" ];
+        path = [ pkgs.iproute ];
+        preStart = ''
+          ip link del dev ${name} 2>/dev/null || true
+          ip link add dev ${name} type wireguard
+        '';
+        serviceConfig.ExecStart = "${pkgs.wireguard}/bin/wg setconf ${name} ${generateConf name values}";
+        postStart = ''
+          ip address add ${values.ip} dev ${name}
+          ip link set up dev ${name}
```

Comment:
The systemd unit is the part I am the less happy with, and also the part I didn't had much time to work on.

I was thinking to split the unit with preStart and postStart and add options to allow users to override the phases if they have more complicated needs like namespace setups.
Do you have any opinion on this point?